packages/compartment-mapper/src/bundle.js

```diff
@@ -0,0 +1,333 @@
+  // index of the corresponding functor.
+  const modulesByKey = {};
+  for (let index = 0; index < modules.length; index += 1) {
+    const module = modules[index];
+    module.index = index;
+    modulesByKey[module.key] = module;
+  }
+  for (const module of modules) {
+    module.indexedImports = Object.fromEntries(
+      Object.entries(module.resolvedImports).map(([importSpecifier, key]) => [
+        importSpecifier,
+        modulesByKey[key].index,
+      ]),
+    );
+  }
+
+  // Only support mjs format.
+  const problems = modules
+    .filter(module => module.parser !== 'premjs')
+    .map(
+      ({ moduleSpecifier, compartmentName, parser }) =>
+        `module ${moduleSpecifier} in compartment ${compartmentName} in language ${parser}`,
+    );
+  if (problems.length) {
+    throw new Error(
+      `Can only bundle applications that only have ESM (.mjs-type) modules, got ${problems.join(
+        ', ',
+      )}`,
+    );
+  }
+
+  const bundle = `\
+(functors => {
+  function cell(name) {
+    const observers = [];
+    let value;
+    function set(newValue) {
+      value = newValue;
+      for (const observe of observers) {
+        observe(value);
+      }
+    }
+    function get() {
+      return value;
+    }
+    function observe(observe) {
+      observers.push(observe);
+      observe(value);
+    }
+    return { get, set, observe, enumerable: true };
+  }
+
+  const cells = [${''.concat(
+    ...modules.map(
+      ({ record: { __fixedExportMap__, __liveExportMap__ } }) => `{
+        ${''.concat(
+          ...Object.keys(__fixedExportMap__).map(
+            exportName => `${exportName}: cell(${q(exportName)}),\n`,
+          ),
+        )}
+        ${''.concat(
+          ...Object.keys(__liveExportMap__).map(
+            exportName => `${exportName}: cell(${q(exportName)}),\n`,
+          ),
+        )}
+      },`,
+    ),
+  )}];
+
+  ${''.concat(
+    ...modules.flatMap(({ index, indexedImports, record: { reexports } }) =>
+      reexports.map(
+        (/* @type {string} */ importSpecifier) => `\
+          Object.defineProperties(cells[${index}], Object.getOwnPropertyDescriptors(cells[${indexedImports[importSpecifier]}]));
+        `,
+      ),
+    ),
+  )}
+
+  const namespaces = cells.map(cells => Object.create(null, cells));
+
+  ${''.concat(
+    ...modules.map(
+      ({
+        index,
+        indexedImports,
+        record: { __liveExportMap__, __fixedExportMap__ },
+      }) => `\
+        functors[${index}]({
+          imports(map) {
+            ${''.concat(
+              ...Object.entries(indexedImports).map(
+                ([importName, importIndex]) => `\
+                  for (const [name, observers] of map.get(${q(
+                    importName,
+                  )}).entries()) {
+                    if (name === '*') {
+                      for (const observer of observers) {
+                        observer(namespaces[${importIndex}]);
```

Comment:
And here, too?  (If so, also the other observer loop just a little lower.). This is smelling like a library function:
```js
const safeApply = (fn, args) => {
   try {
     fn(...args);
   } catch (e) {
     // at the very least, ignore.
   }
};
```

Reply:
This is a good point, and it led me to add an explicit reference error in the case where the imported cell isn’t defined, since that will generally be a better user experience.

Tho, there are no third-party observer functions; they’re all the same cell observer inlined above. If that’s broken, I definitely want a runtime exception to bubble up and report the line and exact nature. I’d do a catch and release to make the error more descriptive, but that level of defensiveness would be hard to apply as a general principle.